api/app/utils/geospatial.py

```diff
@@ -0,0 +1,165 @@
+
+
+def get_geospatial_metadata(data_source, options: Optional[GeospatialOptions]):
+    """
+    Extracts metadata defined by the options parameter from the data_source raster
+
+    :param data_source: input raster to extract metadata from
+    :param options: defines the metadata the caller is interested in extracting from the raster
+    :return: optional metadata of geotransform, project, x and y sizes from the raster
+    """
+    geotransform = None
+    projection = None
+    x_size = None
+    y_size = None
+    if options is None:
+        return (geotransform, projection.x_size, y_size)
+
+    if options.include_geotransform:
+        geotransform = data_source.GetGeoTransform()
+
+    if options.include_projection:
+        projection = data_source.GetProjection()
+
+    if options.include_x_size:
+        x_size = data_source.RasterXSize
+
+    if options.include_y_size:
+        y_size = data_source.RasterYSize
+
+    return (geotransform, projection, x_size, y_size)
+
+
+def read_raster_data(data_source):
+    """
+    Read raster data and return as array from the data source
+
+    :param data_source: data source to read the data from
+    :return: raster data as array
+    """
+    data_band = data_source.GetRasterBand(1)
+    data_array = data_band.ReadAsArray()
+    return data_array
+
+
+def get_raster_data(data_source, options: Optional[GeospatialOptions]) -> GeospatialReadResult:
+    """
+    Returns the raster data and optional metadata the caller is interested defined by the options parameter
+
+    :param data_source: raster data source
+    :param options: options defining the metadata the caller is interested in
+    :return: raster data array and potential metadata
+    """
+    (data_geotransform, data_projection, data_x_size, data_y_size) = get_geospatial_metadata(data_source, options)
+    data_array = read_raster_data(data_source)
+    data_source = None
+    del data_source
+    return GeospatialReadResult(data_array=data_array, data_geotransform=data_geotransform, data_projection=data_projection, data_x_size=data_x_size, data_y_size=data_y_size)
+
+
+def _get_raster_data_source(key: str, s3_data, options: Optional[GeospatialOptions]):
+    """
+    Retrieves geospatial file from S3. If the options include Vector2RasterOptions, it will transform
+    the vector into a raster and return the raster datasource. If no Vector2RasterOptions are included,
+    it will assume the file is a raster and try to read it as a raster. The caller is expected to know whether
+    the geospatial file is a raster or vector and supply the appropriate options.
+
+    :param key: the object store key pointing to the desired geospatial file
+    :param s3_data: the object store content blob
+    :param options: options defining the desired metadata as well as vector options if the file is a vector
+    :return: raster data source
+    """
+    mem_path = f"/vsimem/{key}"
+    gdal.FileFromMemBuffer(mem_path, s3_data)
```

Comment:
Is this line needed? It looks like we open the mem_path here, in the `if` block with `gdal.OpenEx` and in the `else` block with `gdal.Open`.

Reply:
Nevermind, I think I figured it out. I was a little confused by the 'dual use' of the key parameter.

Reply:
Hmm maybe I should create separate `vsimem` paths to avoid the confusion

Reply:
I adjusted the code to use separate path variables for in memory s3 blob and in memory generated raster here: https://github.com/bcgov/wps/pull/3800/commits/7e10bbd8b280a8d1abc84b66a30a863c98d75126